### PR TITLE
[FIL-385] WeakPtr cache problem

### PR DESCRIPTION
**Pull request by ortyomka**
Signed-off-by: ortyomka <iurin.art@gmail.com>

### CODE REVIEW

File: core/primitives/tipset/load.hpp [hunk 1]
```diff
@@ -5,42 +5,85 @@
...
+    boost::optional<TipsetCPtr> getFromCache(uint64_t index,
+                                             const TipsetKey &key);
+    TipsetCPtr getFromCache(uint64_t index);
+
     TsLoadPtr ts_load;
+
     std::mutex mutex;
-    boost::compute::detail::lru_cache<TipsetKey, TipsetCPtr> cache;
+    std::map<TipsetKey, uint64_t> map_cache;
+    std::vector<std::shared_ptr<LoadNode>> tipset_cache;
```
Comment (turuslan) [inline core/primitives/tipset/load.hpp:84]:
```suggestion
    std::vector<LoadNode> tipset_cache;
```

File: core/primitives/tipset/load.hpp [hunk 2]
```diff
@@ -5,42 +5,85 @@
...
+#include <map>
 #include <mutex>
 
 #include "primitives/tipset/tipset.hpp"
 
 namespace fc::primitives::tipset {
   struct TsLazy {
     TipsetKey key;
-    TsWeak weak{};
+    uint64_t cache_index;
```
Comment (turuslan) [inline core/primitives/tipset/load.hpp:16]:
```suggestion
    uint64_t cache_index{};
```

Comment (turuslan) [reply core/primitives/tipset/load.hpp:16]: `using` alias would be good

File: core/primitives/tipset/load.cpp [hunk 3]
```diff
@@ -45,26 +33,156 @@ namespace fc::primitives::tipset {
...
+  }
+
+  uint64_t TsLoadCache::cacheInsert(TipsetCPtr tipset) {
+    std::lock_guard lock{mutex};
+
+    auto res = map_cache.insert(std::make_pair(tipset->key, 0));
+
+    if (tipset_cache.empty()) {
+      begin_index = 0;
+      end_index = 1;
```
Comment (turuslan) [inline core/primitives/tipset/load.cpp:86]: if "end" = "last" then start with 0.

File: core/primitives/tipset/load.cpp [hunk 4]
```diff
@@ -45,26 +33,156 @@ namespace fc::primitives::tipset {
...
+      tipset_cache[end_index - 1]->next = end_index;
+
+      auto tcn{std::make_shared<LoadNode>()};
+      tcn->tipset = std::move(tipset);
+      tcn->prev = end_index - 1;
+      tcn->next = end_index;
+      tcn->it = res.first;
+      tcn->it->second = end_index;
+      tipset_cache.push_back(tcn);
+      return end_index++;
```
Comment (turuslan) [inline core/primitives/tipset/load.cpp:107]: order seems wrong

**Review by turuslan — APPROVED** [review]

**Review by Alexey-N-Chernyshov — APPROVED** [review]

**Pull request merged**